OpenNefia.Content/CharaMake/CharaMakeAttributeRerollLayer.cs

```diff
@@ -168,10 +172,35 @@ public override void Initialize(CharaMakeData args)
             IsInitialized = true;
         }
 
-        private void Reroll()
+        private void Reroll(bool playSound = true)
         {
-            Sounds.Play(Protos.Sound.Dice);
-            //TODO add rerolling
+            if (playSound)
+                Sounds.Play(Protos.Sound.Dice);
+
+            if (!Data.TryGetValue("race", out RacePrototype race))
+            {
+                Logger.WarningS("charamake", "no race prototype in charmake data");
+                return;
+            }
+            if (!Data.TryGetValue("class", out ClassPrototype @class))
+            {
+                Logger.WarningS("charamake", "no class prototype in charmake data");
+                return;
+            }
+
+            foreach (var item in List)
+            {
+                if (item.Data is not AttributeRerollData.Attribute attr)
+                    continue;
+                if (attr.Locked)
+                    continue;
+
+                race.BaseSkills.TryGetValue(attr.Id, out var level);
+                if (@class.BaseSkills.TryGetValue(attr.Id, out var classLevel))
+                    level += classLevel;
+
+                attr.Amount = new Random().Next((level / 2) + 1, level + 1);
```

Comment:
We have an `IRandom` class already, which you can use with:

```csharp
[Dependency] private readonly IRandom _random = default!;
```

This will be used for substituting HSP's random generator where backwards compatibility is necessary, like title generation.

Reply:
changed in 24f46e9